lib/maths/CDataFrameUtils.cc

```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+
+    std::size_t numberSamples{std::min(NUMBER_SAMPLES_TO_COMPUTE_MIC, frame.numberRows())};
+
+    auto method = frame.inMainMemory() ? categoryMicWithColumnDataFrameInMemory
+                                       : categoryMicWithColumnDataFrameOnDisk;
+
+    TSizeDoublePrVecVec mics(method(numberThreads, frame, columnMask, targetColumn,
+                                    numberSamples, minimumFrequency));
+
+    for (auto& categoryMics : mics) {
+        std::sort(categoryMics.begin(), categoryMics.end(),
+                  [](const TSizeDoublePr& lhs, const TSizeDoublePr& rhs) {
+                      return COrderings::lexicographical_compare(
+                          -lhs.second, lhs.first, -rhs.second, rhs.first);
+                  });
+    }
+
+    return mics;
+}
+
 CDataFrameUtils::TDoubleVec CDataFrameUtils::micWithColumn(const core::CDataFrame& frame,
-                                                           const TSizeVec& columnMask,
+                                                           TSizeVec columnMask,
                                                            std::size_t targetColumn) {
 
+    if (targetColumn >= frame.numberColumns()) {
+        HANDLE_FATAL(<< "Internal error: target column out of bounds '"
+                     << targetColumn << " >= " << frame.numberColumns()
+                     << "'. Please report this problem.");
+        return TDoubleVec(frame.numberColumns(), 0.0);
+    }
+
+    const auto& columnIsCategorical = frame.columnIsCategorical();
+    columnMask.erase(std::remove_if(columnMask.begin(), columnMask.end(),
+                                    [&columnIsCategorical](std::size_t i) {
+                                        return columnIsCategorical[i];
+                                    }),
+                     columnMask.end());
+
     std::size_t numberSamples{std::min(NUMBER_SAMPLES_TO_COMPUTE_MIC, frame.numberRows())};
 
-    return frame.inMainMemory()
-               ? micWithColumnDataFrameInMainMemory(numberSamples, frame, columnMask, targetColumn)
-               : micWithColumnDataFrameOnDisk(numberSamples, frame, columnMask, targetColumn);
+    auto method = frame.inMainMemory() ? micWithColumnDataFrameInMemory
+                                       : micWithColumnDataFrameOnDisk;
+
+    return method(frame, columnMask, targetColumn, numberSamples);
 }
 
 bool CDataFrameUtils::isMissing(double x) {
     return CMathsFuncs::isFinite(x) == false;
 }
 
-CDataFrameUtils::TDoubleVec
-CDataFrameUtils::micWithColumnDataFrameInMainMemory(std::size_t numberSamples,
-                                                    const core::CDataFrame& frame,
-                                                    const TSizeVec& columnMask,
-                                                    std::size_t targetColumn) {
+CDataFrameUtils::TSizeDoublePrVecVec
+CDataFrameUtils::categoryMicWithColumnDataFrameInMemory(std::size_t numberThreads,
```

Comment:
nit: I would arrange the lines in `categoryMicWithColumnDataFrameInMemory` and `categoryMicWithColumnDataFrameOnDisk` in a way that they correspond as much as possible. It may simplify comparison, bug search, or refactoring in the future.

Reply:
I tweaked the function layouts in [b0f154e](https://github.com/elastic/ml-cpp/pull/539/commits/b0f154e35b38c7d66ee559a8a24cc668da8805af). (And also `micWithColumnDataFrameInMemory` and `micWithColumnDataFrameOnDisk`.)